RockSniffer/RPC/AlbumArtResolver.cs

```diff
@@ -0,0 +1,110 @@
+                    {
+                        //Console.WriteLine($"cache miss on {key}");
+                        if (GetFromAppleMusic(songInfo) is (string, string) validResult)
+                        {
+                            cache[key] = validResult;
+                        }
+                    }
+
+            return cache.Keys.Contains(key) ? cache[key] : (null, $"{key}");
+        }
+
+        /// <summary>
+        /// Attempt to obtain album cover from Apple Music
+        /// </summary>
+        /// <param name="songInfo">Details about the given song</param>
+        protected (string URL, string DisplayText)? GetFromAppleMusic(SongDetails songInfo)
+        {
+            HttpClient httpClient = new HttpClient();
+            httpClient.DefaultRequestHeaders.Add("User-Agent", "RockSniffer");
+
+            string baseUrl = "https://itunes.apple.com/search";
+            NameValueCollection queryString = System.Web.HttpUtility.ParseQueryString(string.Empty);
+            queryString.Add("media", "music");
+            queryString.Add("limit", "25");
+            queryString.Add("term", $"{songInfo.artistName} {songInfo.songName}");
+
+            var webRequest = new HttpRequestMessage(HttpMethod.Get, baseUrl + "?" + queryString.ToString());
+
+            using (var reader = new StreamReader(httpClient.Send(webRequest).Content.ReadAsStream()))
+            {
+                if (JsonConvert.DeserializeObject<AppleRequestResponse>(reader.ReadToEnd()) is AppleRequestResponse response)
+                {
+                    AppleResult? bestResult = null;
+
+                    // Attempt to find a full match, or at least valid result
+                    foreach (AppleResult appleResult in response.results)
+                    {
+                        if (appleResult.artistName != null && appleResult.collectionName != null && appleResult.trackName != null)
+                        {
+                            if (bestResult == null)
+                                bestResult = appleResult;
+
+                            if ((appleResult.artistName.Contains(songInfo.artistName)
+                                 || songInfo.artistName.Contains(appleResult.artistName))
+                                && (appleResult.collectionName.Contains(songInfo.albumName)
+                                    || songInfo.albumName.Contains(appleResult.collectionName)))
+                            {
+                                bestResult = appleResult;
+                                break;
+                            }
+                        }
+                    }
+
+                    // If bestResult isn't null (is valid result, perhaps even full match), return the URL and description
+                    if (bestResult is AppleResult appleResult1)
+                    {
+                        return (bestResult.artworkUrl100, $"{bestResult.artistName} - {bestResult.collectionName} (fetched from Apple Music)");
```

Comment:
Clarify that the cover image was fetched from Apple Music?

Reply:
Also consider the maximum length of the field on Discord RPC. Do long artist / album names need to be truncated?

Reply:
Okay, that's a good point, I can see how one could read that that way. The truncation I just fixed in the new commit 😄